rules/CodeQuality/Rector/FunctionLike/SimplifyUselessLastVariableAssignRector.php

```diff
@@ -0,0 +1,288 @@
+            }
+
+            $assignNode = $stmt->expr;
+            if (! $assignNode instanceof Assign && ! $assignNode instanceof AssignOp) {
+                continue;
+            }
+
+            $currentVariableNode = $assignNode->var;
+            if (! $currentVariableNode instanceof Variable) {
+                continue;
+            }
+
+            if ($this->nodeNameResolver->areNamesEqual($returnVariable, $currentVariableNode)) {
+                return $stmt;
+            }
+        }
+
+        return null;
+    }
+
+    private function shouldSkipOnAssignStmt(Expression $assignStmt): bool
+    {
+        if ($this->hasSomeComment($assignStmt)) {
+            return true;
+        }
+
+        $assign = $assignStmt->expr;
+        if (! $assign instanceof Assign && ! $assign instanceof AssignOp) {
+            return true;
+        }
+
+        $variable = $assign->var;
+        if (! $variable instanceof Variable) {
+            return true;
+        }
+
+        $value = $assign->expr;
+        if ($this->exprAnalyzer->isDynamicExpr($value)) {
+            return true;
+        }
+
+        if ($value instanceof Array_ && $this->arrayManipulator->isDynamicArray($value)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private function hasSomeComment(Expression $stmt): bool
+    {
+        if ($stmt->getComments() !== []) {
+            return true;
+        }
+
+        return $stmt->getDocComment() !== null;
+    }
+
+    /**
+     * @param Stmt[] $stmts
+     */
+    private function isAssigmentUseless(array $stmts, Expression $assignStmt, Return_ $return): bool
+    {
+        $assign = $assignStmt->expr;
+        if (! $assign instanceof Assign && ! $assign instanceof AssignOp) {
+            return false;
+        }
+
+        $variable = $assign->var;
+        if (! $variable instanceof Variable) {
+            return false;
+        }
+
+        //Find the variable usage
+        $variableUsageNodes = $this->betterNodeFinder->find(
+            $this->getNodesInRange($stmts, $assignStmt, $return),
+            fn (Node $node): bool => $this->exprUsedInNodeAnalyzer->isUsed($node, $variable)
+        );
+
+        //Should be exactly used 2 times (assignment + return)
+        return count($variableUsageNodes) !== 2;
+    }
+
+    /**
+     * @param Stmt[] $stmts
+     * @return Stmt[]
+     */
+    private function getNodesInRange(array $stmts, Expression $start, Return_ $end): array
+    {
+        $resultStmts = [];
+        $wasStarted = false;
+
+        foreach ($stmts as $stmt) {
+            if ($stmt === $start) {
+                $wasStarted = true;
+            }
+
+            if ($wasStarted) {
+                $resultStmts[] = $stmt;
+            }
+
+            if ($stmt === $end) {
+                return $resultStmts;
+            }
+        }
+
+        if ($wasStarted) {
+            throw new ShouldNotHappenException('End node was not found.');
+        }
```

Comment:
I prefer to not throw Exception in rector rule when possible, should just skip the refactor instead.

Reply:
This exception only occur on misusage of the private method. Throwing an exception for this use-case, should be the preferred way, or not?

Hiding the exception cause an unexpected behavior by skipping the rule. The next question of the developer would be, why the rule is skipped after my code change, and he needs to debug. With an exception here, it's clear for a developer what the problem is.

An alternative solution could be the usage of `assert` and skipping the rule.

Reply:
private method should not has possible missusage, as it scoped.

The method that call private method should has unit test cover it if any.

Reply:
I removed the exception.